### DISCUSSION TITLE
Porespy is ploting a pore sheet closer to the PNM. How to resolve this?

### DISCUSSION THREAD

**Discussion by lucianoguedes8** [PMEAL/porespy#994] [Q&A]
Hello!
I'm geologist and I'm trying to make PNMs. I used PoreSpy code to generate a network, but after i use PoreSpy library codes, after I creating the PNM, when I try to plot there is a sheet made of porebodies on my model. 

I believe this pore sheet do not came from my segmented image, because I removed some lateral slices from my image stack and this didn't remove this problem.

To generate my network, I created a matrix maded of a image stack with Numpy library. The image stack is maded by image slices.
So, I have 3 important variables that I worked before use PoreSpy codes:
image_slice
image_stack  
stack_matrix → Furthermore, I have made variables named by new_stack_matrix and pnm_stack_matrix

Can someone help me to solve this? Thank you all.

Print screen of two attempts. The second, I have removed some slices from the side of image_stack.

![image](https://github.com/user-attachments/assets/9146258e-38e5-491c-b207-cfdc4be2b25d)
![image](https://github.com/user-attachments/assets/b740de73-9213-41a1-9781-d92d3bcebeb9)

My code after other codes from image uploads:

"""Generating Network"""

# Generate seeds (I don't believe this will help)
ps.visualization.set_mpl_style()
np.random.seed(10)

# Applying SNOW algorithm - Sub-Network of an Over-segmented Watershed
snow_output = ps.networks.snow2(pnm_stack_matrix, voxel_size=1)

# Generate the configuration of the tridimensional network by PoreSpy library codes: 
pn = op.io.network_from_porespy(snow_output.network)

# Exhibit network configurations:
print(pn)

Now, I expected that I can visualize the network, but this sheet appears at the side of my model.
There is a command that I use an "image slice", I believe this could generate the pores.

# Plotting the network:
fig, ax = plt.subplots(figsize=[8, 8])
ax.imshow(image_slice.T, cmap=plt.cm.bone);
op.visualization.plot_coordinates(ax=fig,
                                  network=pn,
                                  size_by=pn["pore.inscribed_diameter"],
                                  color_by=pn["pore.inscribed_diameter"],
                                  markersize=200)
op.visualization.plot_connections(network=pn, ax=fig)
ax.axis("on");

**Comment by jgostick** [comment]
check your network to see which pores all have identical x coordinates.  Then you can delete them from the network.  I would assume the problem is coming from your original image, but you say that you've tried trimming the image.  Nonetheless, I don't see how else they could appear there.

**Reply by lucianoguedes8** [reply]
All right. Thank you, professor.

I'll see it and I write an answer about this soon.

**Reply by lucianoguedes8** [reply]
Sorry, professor.

I made another issue in github. I could not check this in octuber of 2024 cuz I was occupied with my undergraduate thesis.

How can I remove this from the PNM generated?

**Reply by jgostick** [reply]
These pores all appear to have the same x-coordinate, and it is the 'maximum' x-value, so:

```python

xmax = pn['pore.coords'][:, 0].max()
hits = pn['pore.coords'][:, 0] = xmax
op.topotools.trim(pn, pores=hits)
```

This *might* work.

**Reply by lucianoguedes8** [reply]
This will affect only the PNM plotting or will affect the PNM Generation too? Because I am trying to make sure that the PNM generated do not have this pores.

In whic part I must put this code? And how can I organize this?
Before PNM generation? After?
Or only in plotting (which I do not want) ?

**Reply by lucianoguedes8** [reply]
Sorry, professor... 

I did this, but this pore sheet was not removed.
But, it seems to be an lateral stack of pores.

After the code that you showed:
![image](https://github.com/user-attachments/assets/b98c0eb4-7fd4-4bbc-9c2e-16989ad6f760)

Before the code that you showed:
![image](https://github.com/user-attachments/assets/061793d0-0435-4c67-8e16-9f43d8bb6be2)

**Reply by jgostick** [reply]
So it seems that those pores are not all at exactly x=xmax. Perhaps try some threshold think like `pn['pore.coords'][:, 0] >= 0.95*xmax`

**Reply by lucianoguedes8** [reply]
Well, I think we made it. But I used the code **pn['pore.coords'][:, 0] >= 0.80*xmax**.

![image](https://github.com/user-attachments/assets/04b516ee-3a4e-49b5-97c9-d758b50db225)

This triangle-like shape of PNM may be associated with perspective of the cutted plot.

I will start the program from zero now and try to use 0.95 of xmax. My concern is if the code will cut the pores from the last part of PNM. Maybe I should use 0.96 or 0.97.

I will send a reply tomorrow or tonight.

**Reply by lucianoguedes8** [reply]
Greetings, professor.

I finished the plotting of my PNM. It seems that this problem was resolved with thresholding X axis.
The triangle-shape of the discussion above is due to an bad execution of the code (everytime that I run the code, I must start it from beggining and I forgot this).
Now, it seems to correct and the problem was resolved:

![image](https://github.com/user-attachments/assets/082da510-9f5a-4ccc-8b89-9bcb08f79425)

Now I will search for means to change the color of the nodes (pore bodies) and the segments (throats) of PNM. I believe I can do it myself, but feel free to bring suggestions on how to accomplish this.

Thanks for the assistance.
